pkg/oc/cli/cmd/registry/login/login.go

```diff
@@ -0,0 +1,309 @@
+		Example: fmt.Sprintf(loginExample, name+" login"),
+		Run: func(c *cobra.Command, args []string) {
+			kcmdutil.CheckErr(o.Complete(f, args))
+			kcmdutil.CheckErr(o.Validate())
+			kcmdutil.CheckErr(o.Run())
+		},
+	}
+
+	flag := cmd.Flags()
+	flag.StringVar(&o.ConfigFile, "to", o.ConfigFile, "The location of the Docker config.json your credentials will be stored in.")
+	flag.StringVarP(&o.ServiceAccount, "service-account", "z", o.ServiceAccount, "Log in as the specified service account name in the specified namespace.")
+	flag.StringVar(&o.HostPort, "registry", o.HostPort, "An alternate domain name and port to use for the registry, defaults to the cluster's configured external hostname.")
+	flag.BoolVar(&o.SkipCheck, "skip-check", o.SkipCheck, "Skip checking the credentials against the registry.")
+
+	return cmd
+}
+
+func (o *LoginOptions) Complete(f *clientcmd.Factory, args []string) error {
+	cfg, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case len(o.ServiceAccount) > 0:
+		ns, _, err := f.DefaultNamespace()
+		if err != nil {
+			return err
+		}
+		client, err := clientset.NewForConfig(cfg)
+		if err != nil {
+			return err
+		}
+		sa, err := client.Core().ServiceAccounts(ns).Get(o.ServiceAccount, metav1.GetOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				return fmt.Errorf("the service account %s does not exist in namespace %s", o.ServiceAccount, ns)
+			}
+			return err
+		}
+		var lastErr error
+		for _, ref := range sa.Secrets {
+			secret, err := client.Core().Secrets(ns).Get(ref.Name, metav1.GetOptions{})
+			if err != nil {
+				lastErr = err
+				continue
+			}
+			if secret.Type != corev1.SecretTypeServiceAccountToken {
+				continue
+			}
+			token := secret.Data[corev1.ServiceAccountTokenKey]
+			if len(token) == 0 {
+				continue
+			}
+			o.Credentials = newCredentials(fmt.Sprintf("system:serviceaccount:%s:%s", ns, o.ServiceAccount), string(token))
+			break
+		}
+		if o.Credentials.Empty() {
+			if lastErr != nil {
+				if kerrors.IsForbidden(lastErr) {
+					return fmt.Errorf("you do not have permission to view service account secrets in namespace %s", ns)
+				}
+				return err
+			}
+			return fmt.Errorf("the service account %s had no valid secrets associated with it", o.ServiceAccount)
+		}
+	default:
+		if len(cfg.BearerToken) == 0 {
+			return fmt.Errorf("no token is currently in use for this session")
+		}
+		o.Credentials = newCredentials("user", cfg.BearerToken)
+	}
+
+	if len(o.HostPort) == 0 {
+		client, err := imageclient.NewForConfig(cfg)
+		if err != nil {
+			return err
+		}
+		ns, _, err := f.DefaultNamespace()
+		if err != nil {
+			return err
+		}
+
+		if registry, internal := findPublicHostname(client, ns, "openshift"); len(registry) > 0 {
+			if ref, err := imageapi.ParseDockerImageReference(registry); err == nil {
+				o.HostPort = ref.Registry
+				if internal {
+					fmt.Fprintf(o.ErrOut, "info: Using internal registry hostname %s\n", o.HostPort)
```

Comment:
If this is info, use `o.Out` not `o.ErrOut`, the latter is only for warning and errors.

Reply:
We should be using stderr for anything that is not directly tied to the output of the command.

Reply:
Info, warning, and error are all stderr.

Reply:
Warning and error -> stderr, if info should be possible to hide use glog with different levels. Any other info is info that should put to stdout just like the rest of the output b/c it's relevant to the user.

Reply:
No, it shouldn’t.  Info is not glog.  Info is information relevant to the user that is not a warning.  It is not normal output.  Stdout is data for processing.  Glog is not human facing the way info is.

Reply:
```
grep -R --include=*.go -E "\"info:" pkg/
pkg//oc/admin/migrate/etcd/ttl.go:	fmt.Fprintf(o.Out, "info: Lease #%d with TTL %d created\n", lease.ID, lease.TTL)
pkg//oc/admin/migrate/etcd/ttl.go:	fmt.Fprintf(o.Out, "info: Attaching lease to %d entries\n", len(objectsResp.Kvs))
pkg//oc/admin/migrate/etcd/ttl.go:		fmt.Fprintf(o.Out, "info: Lease already attached to %d entries, no change made\n", alreadyAttached)
pkg//oc/admin/migrate/migrator.go:		fmt.Fprintf(out, "info: to rerun only failing resources, add --include=%s\n", strings.Join(t.resourcesWithErrors.List(), ","))
pkg//oc/admin/router/router.go:			fmt.Fprintf(errout, "info: password for stats user %s has been set to %s\n", cfg.StatsUsername, cfg.StatsPassword)
pkg//oc/cli/cmd/image/mirror/mirror.go:				fmt.Fprintf(o.ErrOut, "info: Filtered all images from %s, skipping\n", src.ref)
pkg//oc/cli/cmd/logs.go:		o.KubeLogOptions.Out.Write([]byte(fmt.Sprintf("info: Logs available at %s\n", urlString)))
pkg//oc/cli/cmd/set/buildhook.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/buildsecret.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/deploymenthook.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/imagelookup.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/probe.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/routebackends.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/triggers.go:			fmt.Fprintf(o.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/volume.go:			fmt.Fprintf(v.Err, "info: %s %q was not changed\n", info.Mapping.Resource, info.Name)
pkg//oc/cli/cmd/set/volume.go:		fmt.Fprintf(v.Err, "info: Generated volume name: %s\n", name)
pkg//oc/cli/util/clientcmd/factory.go:		glog.V(1).Infof("info: the output version specified is invalid. %s\n", defaultVersionInfo)
```

Reply:
All glog output goes to stderr, but `info:` is effectively v(0) with the prefix (which is not useful for a human interacting with a cli).  The user *should* have the information.  The user *should not* have to filter that out, i.e.:

```
$ curl https://$( oc registry info )
```

shouldn't send `info: The registry is internal` to curl - it should print that on stderr so the shell captures only the url, not the output

Reply:
The TTL command is outputting to stdout because it prints nothing else.